### ISSUE TITLE
Improve Formly config structure

### ISSUE THREAD

**Issue by aitboudad** [ngx-formly/ngx-formly#2853]
This is part of #2571, I think we need to improve our Formly config for better readability (avoid long JSON), easy to remember, and quick typing.

1- `templateOptions` => `ui` (or `props` :thinking:):

```ts
{
  key: 'name',
  type: 'input',
  ui: { label: 'Name' },
  expressions: {
    'ui.disabled': 'model.disabled'
  }
}
```

2- `expressionProperties|hideExpression` => `expressions`:

```ts
{
  key: 'name',
  type: 'input',
  expressions: {
    hide: 'model.disabled',
  }
}
```

3- `fieldGroup` => `group`:

```ts
{
  key: 'address',
  group: [ ... ]
}
```

4- `fieldArray` => `array`:

```ts
{
  key: 'addresses',
  array: { ... }
}
```

let me know if you've more suggestions!

**Comment by nthonymiller** [comment]
@aitboudad -  Some suggestions:
1 - Template Options, prefer the naming props over ui.

1 and 2 - Seems confusing the naming of expressionProperties (in 1) and expressions (in 2).

4 - Array
      1. Specify either a component (via a specific form field) or a custom form group to be rendered. 
      2. Ability to use a wrapper component to handle the adding and deleting of items for the array, reasoning behind this currently need to have this all in a single component.
   
Personally, I've struggled with the usage of hideExpression, I guess my mind works better on when an item is visible rather than invisible.

**Comment by nirmpate** [comment]
1 - I am leaning more towards naming it "props" than "ui".  I use templateOptions to add extra properties to my components and usually, I don't reference them in my HTML file.  I mainly use them as props when I initiate the component. 

2- This is great.  I am for grouping all the expressions in an object called expressions.

**Comment by beeman** [comment]
I really like the proposal in general @aitboudad! 

Personally I like `ui` over `props`, but I won't have any objections to the latter either.

Overall, great improvement!

**Comment by manuelmeister** [comment]
This would be a huge breaking change. How can we mitigate migration?

While I think the changes make sense, I'm afraid it would require too many code changes in our team and we probably won't be able to upgrade as a result.
So we would be missing out on the cool improvements like fieldChanges, better typing of `Field.to`, …

**Comment by MaxKless** [comment]
Maybe a custom schematic or migration script with something like [ts-morph](https://ts-morph.com/) could ease the migration.

**Comment by aitboudad** [comment]
@manuelmeister the old syntax will be supported in `v6` (no breaking change :warning:) which will be there for at least one year and maybe we also could provide an extension for the legacy syntax after `v6` (if requested).

@MaxKless schematic is the right way for us I think :thinking:

**Comment by manuelmeister** [comment]
I think this is only a problem because the documentation/reference is not very clear. I often had to resort to reading the formly source code to fully understand and use the many features it provides.

I think the new [Angular docs](https://angular.io/docs) are not bad at explaining the various concepts and their implications.

**Comment by MaxKless** [comment]
Honestly I'm not sure I would rename everything either. I agree that 'props' is a better name than 'templateOptions' - I use it as @nirmpate describes too so some templateOptions don't even end up in the template. 

Overall, I think @FritzHerbers raises a lot of good point so I just want to restate that I also feel like it would cause more trouble than it's worth.

**Comment by GlauberF** [comment]
I created using ngx formly a CMS, that is, the formly structure is in the database, if I need to rename it would be a lot of work.
I believe this is not something that would necessarily need to change.
However, moving, would have to continue working in other versions.

In the case of item 4, using the word array, a reserved word, would not be a good choice.

**Comment by manuelmeister** [comment]
1. I would go for `props` because they are just properties you can set
2. perfect
3. group is sensible
4. I would go with something like `arrayTemplate`, as it is not actually an array but the template for each item and is then inserted into the fieldGroup.

I think this would be still a big win for formly, as the new names are more understandable. The old names would just be deprecated and eventually removed with v7.

**Comment by aitboudad** [comment]
Based on your feedback I'm going to implement the `props` and `expressions` only, later we may check others:

1- `templateOptions` => `props`:

```ts
{
  key: 'name',
  type: 'input',
  props: { label: 'Name' },
}
```

2- `expressionProperties|hideExpression` => `expressions`:

```ts
{
  key: 'name',
  type: 'input',
  expressions: {
    hide: 'model.disabled',
  }
}
```

*Note*: The old syntax will be supported in v6

**Comment by xunjianxiang** [comment]
Personally I like children over group

**Comment by aitboudad** [comment]
left `fieldGroup` and `fieldArray`, for `fieldGroup` I think I've got a better name which is `items`:

```patch
{
- fieldGroup: [...],
+ items: [...]
}
```

I leave this for vote :+1:  or :-1:!

**Comment by kenisteward** [comment]
@aitboudad 

I absolutely love items over field array.

But that's mostly because I'm about that json-schema life :) makes the translation easier!

Which btw we will need a rework update for the translation layer for that right?

**Comment by manuelmeister** [comment]
Or fieldGroup => subfields
and fieldArray => arrayTemplate

**Comment by kenisteward** [comment]
> left `fieldGroup` and `fieldArray`, for `fieldGroup` I think I've got a better name which is `items`:
> 
> ```patch
> {
> - fieldGroup: [...],
> + items: [...]
> }
> ```
> 
> I leave this for vote :+1:  or :-1:!

Also you did mean group = fieldGroup and items=fieldArray just so I understand right?

**Comment by aitboudad** [comment]
> Also you did mean group = fieldGroup and items=fieldArray just so I understand right?

@kenisteward `items = fieldGroup`

**Comment by FritzHerbers** [comment]
For fieldGroup, I would at first prefer: `group`, as I can project a record or group fields for layouting.
But as it can logically contain a group (array of fields) or an array of groups (array of fields = rows, with each of them having items = columns), `items` would fit both views.

fieldArray, as it is a template, as was suggested `arrayTemplate`. 
Derived that it is a template for the "items" of an array: `arrayItemsTemplate`, suggests imo better that it is a template for the items property when you want to use it as an array.

**Comment by MaxKless** [comment]
I'm not sure about items. I get that it's generic for both field groups and arrays but at least for field groups it doesn't really communicate the fact that the `items` are just more field definitions. But that's a minor thing so it might not matter in the grand scheme of things.

Btw, if you're already enhancing the `FormlyFieldConfig` type, do you think there is any way to make it extendable without sacrificing type safety. For example, if I have some special properties that I would like to consume in an extension or similar, right now I'm sort of forced to put in under `templateOptions`. Have you thought about this?

**Comment by aitboudad** [comment]
> I'm not sure about items. I get that it's generic for both field groups and arrays but at least for field groups it doesn't really communicate the fact that the `items` are just more field definitions. But that's a minor thing so it might not matter in the grand scheme of things.

The main reason of considering `items` is its generic and represent the field items (children) regardless its a `group`, an `array` or just a field content.

For now I'll let things as it is which give us some time to see how things go with `props` and `expressions`, later probably I'd go with `items` if we didn't came up with a better alternative.

> Btw, if you're already enhancing the `FormlyFieldConfig` type, do you think there is any way to make it extendable without sacrificing type safety. For example, if I have some special properties that I would like to consume in an extension or similar, right now I'm sort of forced to put in under `templateOptions`. Have you thought about this?

`props` is just an alias of `templateOptions`, so it should work in both direction but I may check if that case is covered as well.

**Comment by manuelmeister** [comment]
>> Btw, if you're already enhancing the FormlyFieldConfig type, do you think there is any way to make it extendable without sacrificing type safety. For example, if I have some special properties that I would like to consume in an extension or similar, right now I'm sort of forced to put in under templateOptions. Have you thought about this?  
>
>props is just an alias of templateOptions, so it should work in both direction but I may check if that case is covered as well.

I think this is gravely needed. This would need more generics in angular module formly config.

Also this is in line with these proposed changes I suggested: https://github.com/ngx-formly/ngx-formly/pull/3242#discussion_r835774902

**Comment by manuelmeister** [comment]
I'm in favor for `items`, if it's explained more extensively in the doc (it can be used as a group, as a subfield or as a template).
But if we go for `items` then I think it would be very important to call `fieldArray` => `arrayItemTemplate`

**Comment by aitboudad** [comment]
This issue has been fixed and released as part of [v6.0.0-beta.0](https://github.com/ngx-formly/ngx-formly/releases/tag/v6.0.0-beta.0) release.

Please let us know, in case you are still encountering a similar issue/problem.
Thank you!
